src/pytorch/pme/pmeCUDA.cu

```diff
@@ -0,0 +1,447 @@
+__device__ void invertBoxVectors(const Accessor<float, 2>& box, float recipBoxVectors[3][3]) {
+    float determinant = box[0][0]*box[1][1]*box[2][2];
+    float scale = 1.0/determinant;
+    recipBoxVectors[0][0] = box[1][1]*box[2][2]*scale;
+    recipBoxVectors[0][1] = 0;
+    recipBoxVectors[0][2] = 0;
+    recipBoxVectors[1][0] = -box[1][0]*box[2][2]*scale;
+    recipBoxVectors[1][1] = box[0][0]*box[2][2]*scale;
+    recipBoxVectors[1][2] = 0;
+    recipBoxVectors[2][0] = (box[1][0]*box[2][1]-box[1][1]*box[2][0])*scale;
+    recipBoxVectors[2][1] = -box[0][0]*box[2][1]*scale;
+    recipBoxVectors[2][2] = box[0][0]*box[1][1]*scale;
+}
+
+__device__ void computeSpline(int atom, const Accessor<float, 2> pos, const Accessor<float, 2> box,
+                          const float recipBoxVectors[3][3], const int gridSize[3], int gridIndex[3], float data[][3],
+                          float ddata[][3], int pmeOrder) {
+    // Find the position relative to the nearest grid point.
+
+    float posInBox[3] = {pos[atom][0], pos[atom][1], pos[atom][2]};
+    for (int i = 2; i >= 0; i--) {
+        float scale = floor(posInBox[i]*recipBoxVectors[i][i]);
+        for (int j = 0; j < 3; j++)
+            posInBox[j] -= scale*box[i][j];
+    }
+    float t[3], dr[3];
+    int ti[3];
+    for (int i = 0; i < 3; i++) {
+        t[i] = posInBox[0]*recipBoxVectors[0][i] + posInBox[1]*recipBoxVectors[1][i] + posInBox[2]*recipBoxVectors[2][i];
+        t[i] = (t[i]-floor(t[i]))*gridSize[i];
+        ti[i] = (int) t[i];
+        dr[i] = t[i]-ti[i];
+        gridIndex[i] = ti[i]%gridSize[i];
+    }
+
+    // Compute the B-spline coefficients.
+
+    float scale = 1.0f/(pmeOrder-1);
+    for (int i = 0; i < 3; i++) {
+        data[pmeOrder-1][i] = 0;
+        data[1][i] = dr[i];
+        data[0][i] = 1-dr[i];
+        for (int j = 3; j < pmeOrder; j++) {
+            float div = 1.0f/(j-1);
+            data[j-1][i] = div*dr[i]*data[j-2][i];
+            for (int k = 1; k < j-1; k++)
+                data[j-k-1][i] = div*((dr[i]+k)*data[j-k-2][i]+(j-k-dr[i])*data[j-k-1][i]);
+            data[0][i] = div*(1-dr[i])*data[0][i];
+        }
+        if (ddata != NULL) {
+            ddata[0][i] = -data[0][i];
+            for (int j = 1; j < pmeOrder; j++)
+                ddata[j][i] = data[j-1][i]-data[j][i];
+        }
+        data[pmeOrder-1][i] = scale*dr[i]*data[pmeOrder-2][i];
+        for (int j = 1; j < pmeOrder-1; j++)
+            data[pmeOrder-j-1][i] = scale*((dr[i]+j)*data[pmeOrder-j-2][i]+(pmeOrder-j-dr[i])*data[pmeOrder-j-1][i]);
+        data[0][i] = scale*(1-dr[i])*data[0][i];
+    }
+}
+
+template <int PME_ORDER>
+__global__ void spreadCharge(const Accessor<float, 2> pos, const Accessor<float, 1> charge, const Accessor<float, 2> box,
+                             Accessor<float, 3> grid, int gridx, int gridy, int gridz, float sqrtCoulomb) {
+    float recipBoxVectors[3][3];
+    invertBoxVectors(box, recipBoxVectors);
+    float data[PME_ORDER][3];
```

Comment:
The number of registers in these kernels is really high. I wonder if computing these in another kernel and fetching from global memory here would be better.